### PR TITLE
Use mutex per worker in ParallelReadBuffer

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Remove unnecessary contention on a single mutex.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/IO/ParallelReadBuffer.cpp [hunk 1]
```diff
@@ -278,7 +277,6 @@ void ParallelReadBuffer::readerThreadFunction(ReadWorkerPtr read_worker)
 
 void ParallelReadBuffer::onBackgroundException()
 {
-    std::lock_guard lock(mutex);
```
Comment (kssenii) [inline src/IO/ParallelReadBuffer.cpp:281]: why we don't need synchronization among all workers here?

Comment (antonio2368) [reply src/IO/ParallelReadBuffer.cpp:281]: nice catch! we should sync

**Comment by antonio2368** [comment]
@kssenii I didn't notice any major improvement with this, but my assumption is that this implementation can handle a higher load because only 2 threads sync at the same time, other threads are free to do reading in background while the first reader is being consumed.
We can also close this and revisit it later on if you think it makes no sense to push it now.

**Comment by kssenii** [comment]
@antonio2368, I think worth merging these changes, but looks like there are some issues:
```
2022.05.03 03:29:12.427963 [ 909 ] {} <Fatal> BaseDaemon: ########################################
2022.05.03 03:29:12.428137 [ 910 ] {} <Fatal> BaseDaemon: ########################################
2022.05.03 03:29:12.428337 [ 909 ] {} <Fatal> BaseDaemon: (version 22.5.1.1, build id: 4A92178FB55BAE52) (from thread 689) (query_id: 54492329-50fb-4476-8b82-abb2be504704) (query: SELECT * FROM url('http://localhost:11111/test/{a,tsv_with_header}.tsv', 'TSV', 'c1 UInt64, c2 UInt64, c3 UInt64')) Received signal Aborted (6)
2022.05.03 03:29:12.428447 [ 910 ] {} <Fatal> BaseDaemon: (version 22.5.1.1, build id: 4A92178FB55BAE52) (from thread 698) (query_id: 54492329-50fb-4476-8b82-abb2be504704) (query: SELECT * FROM url('http://localhost:11111/test/{a,tsv_with_header}.tsv', 'TSV', 'c1 UInt64, c2 UInt64, c3 UInt64')) Received signal Aborted (6)
2022.05.03 03:29:12.428528 [ 909 ] {} <Fatal> BaseDaemon: 
2022.05.03 03:29:12.428674 [ 910 ] {} <Fatal> BaseDaemon: 
2022.05.03 03:29:12.428730 [ 909 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f80fd48d03b 0x7f80fd46c859 0x301392b7 0x301d9a2a 0x17bb00ca 0x279e2cf2 0x17ba37c9 0x17c1799a 0x28608b53 0x28608ad7 0x28608a50 0x286089fc 0x28608977 0x286088ff 0x28433628 0x28430bf7 0x2842f653 0x2843682a 0x28436753 0x28436592 0x284363ed 0x284362b5 0x2843627d 0x28436255 0x28436220 0x17be0b46 0x17bdd075 0x17c8e46f 0x17c94e84 0x17c94e15 0x17c94d45 0x17c94682 0x7f80fd644609 0x7f80fd569163
2022.05.03 03:29:12.428916 [ 910 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f80fd48d03b 0x7f80fd46c859 0x301392b7 0x301d9a2a 0x17bb00ca 0x279e2cf2 0x17ba37c9 0x17c1799a 0x28608b53 0x28608ad7 0x28608a50 0x286089fc 0x28608977 0x286088ff 0x28433628 0x28430bf7 0x2842f653 0x2843682a 0x28436753 0x28436592 0x284363ed 0x284362b5 0x2843627d 0x28436255 0x28436220 0x17be0b46 0x17bdd075 0x17c8e46f 0x17c94e84 0x17c94e15 0x17c94d45 0x17c94682 0x7f80fd644609 0x7f80fd569163
2022.05.03 03:29:12.428980 [ 909 ] {} <Fatal> BaseDaemon: 4. raise @ 0x7f80fd48d03b in ?
2022.05.03 03:29:12.429174 [ 910 ] {} <Fatal> BaseDaemon: 4. raise @ 0x7f80fd48d03b in ?
2022.05.03 03:29:12.429198 [ 909 ] {} <Fatal> BaseDaemon: 5. abort @ 0x7f80fd46c859 in ?
2022.05.03 03:29:12.429319 [ 910 ] {} <Fatal> BaseDaemon: 5. abort @ 0x7f80fd46c859 in ?
2022.05.03 03:29:12.527693 [ 910 ] {} <Fatal> BaseDaemon: 6. /build/build_docker/../contrib/libcxx/src/debug.cpp:0: std::__1::__libcpp_abort_debug_function(std::__1::__libcpp_debug_info const&) @ 0x301392b7 in /usr/bin/clickhouse
2022.05.03 03:29:12.528417 [ 909 ] {} <Fatal> BaseDaemon: 6. /build/build_docker/../contrib/libcxx/src/debug.cpp:0: std::__1::__libcpp_abort_debug_function(std::__1::__libcpp_debug_info const&) @ 0x301392b7 in /usr/bin/clickhouse
2022.05.03 03:29:12.616334 [ 910 ] {} <Fatal> BaseDaemon: 7. /build/build_docker/../contrib/libcxx/src/mutex.cpp:49: std::__1::mutex::unlock() @ 0x301d9a2a in /usr/bin/clickhouse
2022.05.03 03:29:12.617420 [ 909 ] {} <Fatal> BaseDaemon: 7. /build/build_docker/../contrib/libcxx/src/mutex.cpp:49: std::__1::mutex::unlock() @ 0x301d9a2a in /usr/bin/clickhouse
2022.05.03 03:29:12.659616 [ 910 ] {} <Fatal> BaseDaemon: 8. /build/build_docker/../contrib/libcxx/include/__mutex_base:142: std::__1::unique_lock<std::__1::mutex>::~unique_lock() @ 0x17bb00ca in /usr/bin/clickhouse
2022.05.03 03:29:12.660950 [ 909 ] {} <Fatal> BaseDaemon: 8. /build/build_docker/../contrib/libcxx/include/__mutex_base:142: std::__1::unique_lock<std::__1::mutex>::~unique_lock() @ 0x17bb00ca in /usr/bin/clickhouse
2022.05.03 03:29:12.773057 [ 910 ] {} <Fatal> BaseDaemon: 9. /build/build_docker/../src/IO/ParallelReadBuffer.cpp:227: DB::ParallelReadBuffer::nextImpl() @ 0x279e2cf2 in /usr/bin/clickhouse
2022.05.03 03:29:12.785559 [ 909 ] {} <Fatal> BaseDaemon: 9. /build/build_docker/../src/IO/ParallelReadBuffer.cpp:227: DB::ParallelReadBuffer::nextImpl() @ 0x279e2cf2 in /usr/bin/clickhouse
2022.05.03 03:29:12.878994 [ 910 ] {} <Fatal> BaseDaemon: 10. /build/build_docker/../src/IO/ReadBuffer.h:85: DB::ReadBuffer::next() @ 0x17ba37c9 in /usr/bin/clickhouse
2022.05.03 03:29:12.920896 [ 909 ] {} <Fatal> BaseDaemon: 10. /build/build_docker/../src/IO/ReadBuffer.h:85: DB::ReadBuffer::next() @ 0x17ba37c9 in /usr/bin/clickhouse
2022.05.03 03:29:12.933296 [ 910 ] {} <Fatal> BaseDaemon: 11.1. inlined from /build/build_docker/../src/IO/ReadBuffer.h:119: DB::ReadBuffer::eof()
2022.05.03 03:29:12.942048 [ 910 ] {} <Fatal> BaseDaemon: 11. ../src/IO/ReadHelpers.cpp:1256: DB::loadAtPosition(DB::ReadBuffer&, DB::Memory<Allocator<false, false> >&, char*&) @ 0x17c1799a in /usr/bin/clickhouse
2022.05.03 03:29:12.995916 [ 909 ] {} <Fatal> BaseDaemon: 11.1. inlined from /build/build_docker/../src/IO/ReadBuffer.h:119: DB::ReadBuffer::eof()
2022.05.03 03:29:12.996276 [ 909 ] {} <Fatal> BaseDaemon: 11. ../src/IO/ReadHelpers.cpp:1256: DB::loadAtPosition(DB::ReadBuffer&, DB::Memory<Allocator<false, false> >&, char*&) @ 0x17c1799a in /usr/bin/clickhouse
2022.05.03 03:29:13.113343 [ 910 ] {} <Fatal> BaseDaemon: 12. /build/build_docker/../src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp:300: DB::fileSegmentationEngineTabSeparatedImpl(DB::ReadBuffer&, DB::Memory<Allocator<false, false> >&, unsigned long, bool, unsigned long) @ 0x28608b53 in /usr/bin/clickhouse
2022.05.03 03:29:13.165170 [ 909 ] {} <Fatal> BaseDaemon: 12. /build/build_docker/../src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp:300: DB::fileSegmentationEngineTabSeparatedImpl(DB::ReadBuffer&, DB::Memory<Allocator<false, false> >&, unsigned long, bool, unsigned long) @ 0x28608b53 in /usr/bin/clickhouse
2022.05.03 03:29:13.281761 [ 910 ] {} <Fatal> BaseDaemon: 13. /build/build_docker/../src/Processors/Formats/Impl/TabSeparatedRowInputFormat.cpp:344: DB::registerFileSegmentationEngineTabSeparated(DB::FormatFactory&)::$_2::operator()(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool) const::'lambda'(DB::ReadBuffer&, DB::Memory<Allocator<false, false> >&, unsigned long)::operator()(DB::ReadBuffer&, DB::Memory<Allocator<false, false> >&, unsigned long) const @ 0x28608ad7 in /usr/bin/clickhouse
```
https://s3.amazonaws.com/clickhouse-test-reports/36801/fd0ba2d10b578717899480f6988af3770e5f2ec6/stateless_tests__debug__actions__[3/3]/runlog.log

**Comment by kssenii** [comment]
> Stateless tests (ubsan, actions) — Timeout, fail: 1

https://github.com/ClickHouse/ClickHouse/issues/36610

> Hung check failed

https://github.com/ClickHouse/ClickHouse/pull/36900

**Review by kssenii — APPROVED** [review]

**Pull request merged**